lib/ash/resource/transformers/get_by_read_actions.ex

```diff
@@ -0,0 +1,241 @@
+      end
+    end)
+  end
+
+  defp transform_action(dsl_state, action) do
+    import Ash.Filter.TemplateHelpers, only: [arg: 1]
+    import Ash.Expr
+
+    action =
+      action.get_by
+      |> Enum.reduce(%{action | get?: true}, fn field, action ->
+        type = type_for_entity(dsl_state, field)
+
+        arguments =
+          if Enum.any?(action.arguments, &(&1.name == field)) do
+            action.arguments
+          else
+            [
+              Transformer.build_entity!(Resource.Dsl, [:actions, :read], :argument,
+                name: field,
+                type: type
+              )
+              | action.arguments
+            ]
+          end
+
+        filter =
+          case action.filter do
+            nil -> expr(ref(^field) == ^arg(field))
+            filter -> where(^filter, ref(^field) == ^arg(field))
+          end
+
+        %{action | arguments: arguments, filter: filter}
+      end)
+
+    {:ok, action}
+  end
+
+  defp type_for_entity(dsl_state, field) do
+    []
+    |> Stream.concat(Transformer.get_entities(dsl_state, [:attributes]))
+    |> Stream.concat(Transformer.get_entities(dsl_state, [:calculations]))
+    |> Stream.concat(Transformer.get_entities(dsl_state, [:aggregates]))
+    |> Enum.find(&(&1.name == field))
+    |> case do
+      aggregate when is_struct(aggregate, Resource.Aggregate) ->
+        {:ok, type} = Resource.Info.aggregate_type(dsl_state, aggregate)
+        Type.get_type(type)
+
+      other ->
+        Type.get_type(other.type)
+    end
+  end
+
+  defp validate_get_by_value(dsl_state, action) do
+    attributes = map_entities(dsl_state, [:attributes], & &1.filterable?)
+    calculations = map_entities(dsl_state, [:calculations], & &1.filterable?)
+    aggregates = map_entities(dsl_state, [:aggregates], & &1.filterable?)
+
+    action.get_by
+    |> Enum.reduce_while(:ok, fn get_by, _ ->
+      cond do
+        Map.has_key?(attributes, get_by) ->
+          if Map.get(attributes, get_by),
+            do: {:cont, :ok},
+            else: {:halt, is_not_filterable_error(dsl_state, action, :attribute, get_by)}
+
+        Map.has_key?(calculations, get_by) ->
+          if Map.get(calculations, get_by),
+            do: {:cont, :ok},
+            else: {:halt, is_not_filterable_error(dsl_state, action, :calculation, get_by)}
+
+        Map.has_key?(aggregates, get_by) ->
+          if Map.get(aggregates, get_by),
+            do: {:cont, :ok},
+            else: {:halt, is_not_filterable_error(dsl_state, action, :aggregate, get_by)}
+
+        true ->
+          {:halt,
+           {:error,
+            dsl_error(
+              dsl_state,
+              [:actions, :read, action.name, :get_by],
+              "`#{inspect(get_by)}` is not a valid attribute, calculation or aggregate"
+            )}}
+      end
+    end)
+  end
+
+  defp validate_existing_arguments(_dsl_state, action) when action.arguments == [], do: :ok
+
+  defp validate_existing_arguments(dsl_state, action) do
```

Comment:
So, its not *necessarily* the case that the types have to match...this kind of thing may be too restrictive. I guess it can't really hurt to start here though.